specs/capella/beacon-chain.md

```diff
@@ -0,0 +1,305 @@
+# Capella -- The Beacon Chain
+
+## Table of contents
+
+<!-- TOC -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+Capella is a consensus-layer upgrade containing a number of features related
+to validator withdrawals. Including:
+* Automatic withdrawals of `withdrawable` validators
+* Partial withdrawals during block proposal
+* Operation to change from `BLS_WITHDRAWAL_PREFIX` to
+  `ETH1_ADDRESS_WITHDRAWAL_PREFIX` versioned withdrawal credentials to enable withdrawals for a validator
+
+## Custom types
+
+## Constants
+
+We define the following Python custom types for type hinting and readability:
+
+| Name | SSZ equivalent | Description |
+| - | - | - |
+| `TransactionType` | `Bytes1` | an EIP-2718 type |
+
+## Preset
+
+### State list lengths
+
+| Name | Value | Unit | Duration |
+| - | - | :-: | :-: |
+| `WITHDRAWAL_TRANSACTION_LIMIT` | `uint64(2**40)` (= 1,099,511,627,776) | withdrawal transactions enqueued in state|
+
+### Execution
+
+| Name | Value | Description |
+| - | - | - |
+| `TX_TYPE_WITHDRAWAL` | `TransactionType('0x05')` | EIP-2718 TX Type |
+| `MAX_WITHDRAWAL_TRANSACTIONS_PER_PAYLOAD` | `uint64(2**4)` (= 16) | Maximum amount of withdrawal transactions allowed in each payload |
+
+## Configuration
+
+## Containers
+
+### Extended Containers
+
+#### `Validator`
+
+```python
+class Validator(Container):
+    pubkey: BLSPubkey
+    withdrawal_credentials: Bytes32  # Commitment to pubkey for withdrawals
+    effective_balance: Gwei  # Balance at stake
+    slashed: boolean
+    # Status epochs
+    activation_eligibility_epoch: Epoch  # When criteria for activation were met
+    activation_epoch: Epoch
+    exit_epoch: Epoch
+    withdrawable_epoch: Epoch  # When validator can withdraw funds
+    withdrawn_epoch: Epoch  # [New in Capella]
```

Comment:
:cry: F s in the chat for Validator container getting +1 depth

Reply:
We don't actually need to do this. We could us another signal to signify withdrawn and wrap it in a function `is_withdrawn`.

e.g. set *all* `*_epoch` to `GENESIS_EPOCH` and then `is_withdrawn` just checks that. Something like that

Reply:
Could it be `withdrawable_epoch <= current_epoch AND effective_balance == 0` (and setting `effective_balance` to 0 could be done in place of setting `withdrawn_epoch` as at current).

Reply:
> e.g. set all *_epoch to GENESIS_EPOCH and then is_withdrawn just checks that. Something like that

That sounds definitely better than the extra field

Reply:
For context, the extra field adds 312500*8 = 2.5MB to a mainnet consensus state, current size is `42.2MB` so it's not insignificant.